Targets

RAtshan opened the pull request:
To complete pipeline II:
1.  I switched 2 object targets to a file target. This required adjustments in the functions arguments as described in the commits. 
2. Then, I switched a file target to a object target. Similar to the above step, I adjusted the function's arguments.

Commit:
Added a `filepath` argument is the `process_data` function. Then changes the return argument to the `filepath`. Implement the changes to `annotate_data` so it can accept the file path from the previous function.

Commit:
Changes the object target argument to file target for `process_data` command. Allowed `annotate_data` to accept a file path instead of object target.

Commit:
Changes `style_data` function to accept a file path and to read a csv file. Then, asked the function to return the csv file.

Commit:
Allowed the `plot_nwis_timeseries` function to read a csv file. This allows the `plot_nwis_timeseries` function to read the csv file that was created by the style_data function.

Commit:
Changes the object target to a file target for the `style_data` function. Then, read the csv file in the `plot_nwis_timeseries` function.

Commit:
Changes the file target `1_fetch/out/site_info.csv` to an object target `site_info`. Please disregard other changes, I was adjusting the different function to find the best file target to change.

Commit:
Changes the `nwis_site_info` function's argument to produce an object target instead of file target.

Commit:
Changes in the `fetch` and `process` functions. These changes are explained in the previous commits.

Commit:
Changes the file target `1_fetch/out/site_info.csv` to an object target `site_info`. Please disregard other changes, I was adjusting the different function to find the best file target to change.

aappling-usgs commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -17,10 +17,10 @@ download_nwis_data <- function(site_nums = c("01427207", "01432160", "01435000",
   site_no <- unique(site_data$site_no)
   site_info <- dataRetrieval::readNWISsite(site_no)
-  write_csv(site_info, fileout)
+  write_csv(site_info) 
```
You've attempted to convert this function so that it returns an object rather than writing a file, but it doesn't work as written. This is what happens when I try to build the associated `site_info` object:
```
> site_info <- scmake('site_info')
Starting build at 2020-06-23 15:21:33
<  MAKE > site_info
[    OK ] site_data
[ BUILD ] site_info                          |  site_info <- nwis_site_info(site_data)
 Error in write_delim(x, path, delim = ",", na = na, append = append, col_names = col_names,  : 
  argument "path" is missing, with no default 
```
Can you fix this? Hint: `write_csv` is an odd thing to be doing in a file that is no longer supposed to be writing a file.

aappling-usgs commented on 2_process/src/process_and_style.R:
```diff
@@ -1,19 +1,23 @@
-annotate_data <- function(site_data_clean, site_filename){
+annotate_data <- function(site_data_clean_path, site_filename){
   site_info <- read_csv(site_filename)
+  site_data_clean <- read_csv(site_data_clean_path)
```
looks just right 👍

aappling-usgs commented on 2_process/src/process_and_style.R:
```diff
@@ -1,19 +1,23 @@
+style_data <- function(style_data_filepath, site_data_annotated){
+  site_data_annotated_out <- read_csv(style_data_filepath)
+  mutate(site_data_annotated_out, station_name = as.factor(station_name))
+  return(style_data_filepath)
```
Before your edits, this function was returning the data.frame result of the `mutate()` call. Now, it's returning the same `style_data_filepath` character string that you passed in as an argument, which isn't the desired result. With the change to file output, this function ought to be writing the output of the `mutate()` call to a file. I'd also just leave off the `return()` call altogether, to help communicate that the important activity in this function is now the creation of the file, not the returning of an R object.

aappling-usgs commented on 3_visualize/src/plot_timeseries.R:
```diff
@@ -1,6 +1,6 @@
 plot_nwis_timeseries <- function(fileout, site_data_styled, width = 12, height = 7, units = 'in'){
   
-  ggplot(data = site_data_styled, aes(x = dateTime, y = water_temperature, color = station_name)) +
+  ggplot(data = read_csv(site_data_styled), aes(x = dateTime, y = water_temperature, color = station_name)) +
```
:+1:

aappling-usgs reviewed (changes requested):
Hi @RAtshan , your changes show that you've learned quite a lot about **scipiper** and file vs object targets :+1:. In addition to some in-line comments below, here are a few other issues to address:

1. It's good practice to avoid committing data files, such as 1_fetch/out/nwis_01427207_data.csv, 2_process/out/site_data_clean, 2_process/out/site_data_clean.csv, and 2_process/out/site_data_styled.csv. This is because data files are relatively large and can use up all the storage space provided by GitHub pretty quickly. We occasionally make intentional exceptions for very small, human-readable data files, but I don't think these 175,000-line files qualify. So, by way of helping you remember this best practice in the future, please remove the data files from your master branch. First delete the data files (or temporarily move them out of your repo, which is the same in git's eyes), then commit and push the deletions. Then, to prevent future issues, add a line to your `.gitignore` file to ignore all files with the `.csv` suffix (use wildcard notation, i.e., `*.csv`). Note that 2_process/out/site_data_clean has no .csv suffix and therefore won't be ignored if you add it back to the repo, so it's probably best to delete it permanently (I assume it was just an accidental output anyway, right?). Remember to git commit the changes to your .gitignore file after you've edited it.
  (  Another note - for real repos, if we accidentally committed a data file, we'd probably go to even greater lengths to get the file completely removed from the repository. If you follow the directions above, the data files will no longer be visible on your HEAD commit on your master branch, but they will still be saved in the history of past changes, so your repo will still be using more storage space than necessary. There are ways to fix this, but they're a pain, so we'll wait until this happens in a real repo to go that extra mile.)

2. I see and appreciate that you changed some variable names to reflect their changed nature (file vs object), and I like that many of the new file argument names also reflect the file's contents (e.g., `site_data_clean_path`, but the naming is still somewhat inconsistent. Specifically, you currently have file arguments called `filepath`, `site_data_clean_path`, `style_data_filepath`, and `site_data_styled`. Can you make them more uniform? There are many "correct" solutions here - for example, the suffix could be "path", "filepath", or "file" as long as it's consistent across the project - so just pick a convention and stick with it. (Note - some inconsistency is inevitable and acceptable in real projects, but naming is super important for readability, and I think it's almost always worth a few extra seconds or minutes to keep things as consistent and informative as you can.)

3. At some point during this course, I think you added the targets `1_fetch/out/nwis_01432160_data.csv`, `1_fetch/out/nwis_01436690_data.csv`, and `1_fetch/out/nwis_01435000_data.csv`...but these won't currently get built if you call `scmake()` without arguments. Can you edit remake.yml so that they are build when you run that command?

4. Once you're pretty sure everything is in place, make sure to run `scmake()` to check that the revised pipeline works. Watch the console messages to see whether the targets you're expecting to run actually do run, and check that the pipeline runs to completion without error.

Commit:
Deleting unnecessary large data files.

Commit:
Added unnecessary `.csv` data files.

Commit:
Renamed the `site_data_styled` file path argument to `site_data_styled_inpath` arguments that I created earlier.

Commit:
Allowed `nwis_site_info` function to take in an path `site_data_outpath` argument to keep track of the csv loaded data file that in the `remake.yml` file.

Commit:
Line 40: Changed the argument name for `process_data` function to `site_data_clean_outpath`. To distinguish it’s the output target.

Line 43: Changed the argument for `annotate_data` function to `site_data_clean_inpath`. This will distinguish it’s the input argument.

Line 46: Changed output target for `styled_data` function.

Line 50: Changed argument name for `site_data_styled` function to  `site_data_styled_inpath`.

Commit:
Line 1: Changed the argument name for `process_data` function to `site_data_clean_outpath`. To distinguish it’s the output target.

Line 5 & 6: Same as above.

Line 19:22 -->
a) The `styled_data` fixed the function,
b) Changed argument name to `style_data_outpath`.
c) Used `data.frame` to ensure the product is a data frame.

Line 9:11 --> Changed the `annotate_data` argument name to `site_data_clean_inpath`.

Commit:
Line 16: building the fetched csv files.

Line 38: changedthe `site_info` target to an object target instead of a file target.

Commit:
Making sure the correct object is retuned form `nwis_site_info` function. Making sure the correct object is retuned form `nwis_site_info` function.

Removing unnecessary `data.frame` command.

aappling-usgs reviewed (approved):
Great!
